src/Chainweb/Version/Utils.hs

```diff
@@ -0,0 +1,434 @@
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+-- |
+-- Module: Chainweb.Version.Utils
+-- Copyright: Copyright © 2020 Kadena LLC.
+-- License: MIT
+-- Maintainer: Lars Kuhtz <lars@kadena.io>
+-- Stability: experimental
+--
+-- Utilities for computing BlockHeights, CutHeights and BlockCounts based on
+-- chainweb versions and chainweb graphs.
+--
+module Chainweb.Version.Utils
+(
+-- * Chain Ids
+  randomChainId
+, someChainId
+
+-- * Chain Graph Properties By Block Height
+, chainGraphs
+, chainGraphsAt
+, lastGraphChange
+, nextGraphChange
+, chainCountAt
+, diameterAt
+, degreeAt
+, chainIdsAt
+, randomChainIdAt
+, someChainIdAt
+, blockCountAt
+, globalBlockCountAt
+, globalBlockRateAt
+, isGraphChange
+
+-- * Chain Graph Properties By Cut Height
+, avgCutHeightAt
+, chainGraphsByCutHeight
+, chainGraphAtCutHeight
+, chainCountAtCutHeight
+, diameterAtCutHeight
+, degreeAtCutHeight
+, blockCountAtCutHeight
+, avgBlockHeightAtCutHeight
+, lastGraphChangeByCutHeight
+, nextGraphChangeByCutHeight
+
+-- * Expected Block and Cut Heights
+, expectedBlockHeightAfterSeconds
+, expectedCutHeightAfterSeconds
+, expectedBlockCountAfterSeconds
+, expectedGlobalBlockCountAfterSeconds
+) where
+
+import Chainweb.BlockHeight
+import Chainweb.Time
+
+import Data.Foldable
+import qualified Data.HashSet as HS
+import qualified Data.Map.Strict as M
+
+import GHC.Stack
+
+import Numeric.Natural
+
+import System.Random
+
+-- internal modules
+
+import Chainweb.Graph
+import Chainweb.Utils
+import Chainweb.Version
+
+-- -------------------------------------------------------------------------- --
+--  Utils
+
+limitHeight :: BlockHeight -> M.Map BlockHeight a -> M.Map BlockHeight a
+limitHeight h = M.takeWhileAntitone (h >=)
+{-# INLINE limitHeight #-}
+
+atHeight :: HasCallStack => BlockHeight -> M.Map BlockHeight a -> a
+atHeight h = snd . fromJuste . M.lookupLE h
```

Comment:
Probably worth documenting unsafety here and `atCutHeight` since the unsafety applies to all the code calling these

Reply:
This is just an internal helper, which is supposed to be called only when it's safe. It's not exported. I hope, the unsafely doesn't leak to any exported function.

I'll add a comment that clarifies it.